### PR TITLE
Fetch disaster and conflict data

### PR DESCRIPTION
**Pull request by AdityaKhatri**

### CODE REVIEW

File: src/views/Gidd/NumberBlock/index.tsx [hunk 1]
```diff
@@ -1,24 +1,41 @@
...
 function NumberBlock({
     label,
+    secondarySubLabel,
     subLabel,
     value,
     className,
     variant = 'normal',
     size = 'small',
+    hideIfNoValue = false,
+    abbreviate = true,
```
Comment (tnagorra) [inline src/views/Gidd/NumberBlock/index.tsx:22]:
```suggestion
    hiddenIfNoValue = false,
    abbreviated = true,
```

File: src/utils/common.ts [hunk 2]
```diff
@@ -28,30 +42,103 @@ export function isValidUrl(url: string | undefined): url is string {
...
-    const maxBounds = bbox(combinedPolygons);
-    return maxBounds as Bounds;
+    ]);
+}
+
+export function removeZero(data?: number) {
+    if (data === 0) {
+        return undefined;
+    }
+    return data;
```
Comment (tnagorra) [inline src/utils/common.ts:128]: Use ternary operator

File: src/utils/common.ts [hunk 3]
```diff
@@ -28,30 +42,103 @@ export function isValidUrl(url: string | undefined): url is string {
...
+        countries: ['BMU', 'CAN', 'USA', 'CHL', 'GUF', 'MTQ', 'GLP', 'SPM', 'SGS', 'BES', 'MSR', 'BLM', 'FLK', 'BVT', 'ABW', 'ATG', 'ARG', 'BHS', 'BRB', 'BLZ', 'BRA', 'CYM', 'COL', 'CRI', 'CUB', 'CUW', 'DMA', 'BOL', 'DOM', 'ECU', 'SLV', 'GRD', 'GTM', 'GUY', 'HTI', 'HND', 'JAM', 'MEX', 'NIC', 'PAN', 'PRY', 'PER', 'PRI', 'KNA', 'LCA', 'VCT', 'MAF', 'TTO', 'TCA', 'SXM', 'URY', 'VIR', 'AIA', 'VGB', 'VEN', 'SUR'],
+    },
+];
+
+export const regionMap = listToMap(regions, (d) => d.key, (d) => d.countries);
+
+export function calcPieSizes(data: { label: string; total: number }[], noOfPies = 5) {
+    const sortedData = data.sort((a, b) => compareNumber(a.total, b.total, -1));
+
+    const topData = [...sortedData];
```
Comment (tnagorra) [inline src/utils/common.ts:106]:
```suggestion
    const sortedData = [...data].sort((a, b) => compareNumber(a.total, b.total, -1));

    const topData = sortedData;
```

File: src/Root/App/index.tsx [hunk 4]
```diff
@@ -1,14 +1,24 @@
...
 import '../../../node_modules/mapbox-gl/dist/mapbox-gl.css';
 import './styles.css';
 
 import Gidd from '../../views/Gidd';
 
 function App() {
+    const requestContextValue = {
+        transformUrl: (d: string) => d,
+        transformOptions: (_: string, options) => processOptions(options),
+    };
```
Comment (tnagorra) [inline src/Root/App/index.tsx:16]: please memoize

File: src/components/CurvedBar/index.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+// NOTE: No types defined by Recharts
+function CustomBar(props: any) {
```
Comment (tnagorra) [inline src/components/CurvedBar/index.tsx:4]:
```suggestion
function CurvedBar(props: any) {
```

File: src/views/Gidd/MapDashboard/index.tsx [hunk 6]
```diff
@@ -25,17 +185,225 @@ function MapDashboard(props: Props) {
...
+    const {
+        pending,
+        response,
+    } = useRequest<MultiResponse<DisplacementData>>({
+        url: 'https://api.idmcdb.org/api/displacement_data',
+        query: {
+            // NOTE: To be changed to 2020 after data is available
+            year: 2020,
+            ci: 'IDMCWSHSOLO009',
+        },
```
Comment (tnagorra) [inline src/views/Gidd/MapDashboard/index.tsx:202]: please memoize

File: src/views/Gidd/MapDashboard/index.tsx [hunk 7]
```diff
@@ -25,17 +185,225 @@ function MapDashboard(props: Props) {
...
+            totalIdpConflictCount: add(
+                definedStockConflict.map((c) => c.conflict_stock_displacement),
+            ) ?? 0,
+            idpConflictCountriesCount: definedStockConflict.length,
+            totalIdpDisasterCount: add(
+                definedStockDisasters.map((c) => c.disaster_stock_displacement),
+            ) ?? 0,
+            idpDisasterCountriesCount: definedStockDisasters.length,
+        };
+    }, [response?.results]);
```
Comment (tnagorra) [inline src/views/Gidd/MapDashboard/index.tsx:255]: why not map and then filter?

File: src/views/Gidd/MapDashboard/index.tsx [hunk 8]
```diff
@@ -25,17 +185,225 @@ function MapDashboard(props: Props) {
...
+        paginatedData,
+        sortedData,
+    } = useMemo(() => {
+        if (!response?.results) {
+            return {
+                paginatedData: [],
+                sortedData: [],
+            };
+        }
+        const finalSortedData = [...response.results].map((d) => ({
```
Comment (tnagorra) [inline src/views/Gidd/MapDashboard/index.tsx:270]: Why create a new array?

File: src/views/Gidd/MapDashboard/index.tsx [hunk 9]
```diff
@@ -25,17 +185,225 @@ function MapDashboard(props: Props) {
...
+                        b[sorting.name],
+                        sorting.direction === 'asc' ? 1 : -1,
+                    );
+                }
+                if (
+                    sorting.name === 'conflict_stock_displacement'
+                    || sorting.name === 'conflict_new_displacements'
+                    || sorting.name === 'disaster_stock_displacement'
+                    || sorting.name === 'disaster_new_displacements'
+                    || sorting.name === 'year'
```
Comment (tnagorra) [inline src/views/Gidd/MapDashboard/index.tsx:291]: Better to use Array.includes

File: src/views/Gidd/MapDashboard/index.tsx [hunk 10]
```diff
@@ -25,17 +185,225 @@ function MapDashboard(props: Props) {
...
+                    return compareNumber(
+                        add([a.disaster_new_displacements, a.conflict_new_displacements]),
+                        add([b.disaster_new_displacements, b.conflict_new_displacements]),
+                        sorting.direction === 'asc' ? 1 : -1,
+                    );
+                }
+                return 1;
+            });
+        }
+        const finalPaginatedData = [...finalSortedData];
```
Comment (tnagorra) [inline src/views/Gidd/MapDashboard/index.tsx:316]: Why make a copy?

File: src/views/Gidd/Disaster/index.tsx [hunk 11]
```diff
@@ -93,44 +169,284 @@ function Disaster(props: Props) {
...
+        if (sorting) {
+            finalPaginatedData.sort((a, b) => {
+                if (
+                    sorting.name === 'geo_name'
+                    || sorting.name === 'event_name'
+                    || sorting.name === 'start_date'
+                    || sorting.name === 'hazard_category'
+                    || sorting.name === 'hazard_sub_category'
+                    || sorting.name === 'hazard_type'
+                    || sorting.name === 'hazard_sub_type'
```
Comment (tnagorra) [inline src/views/Gidd/Disaster/index.tsx:314]: Better to use array.includes method

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/Gidd/MapDashboard/index.tsx (hunk 7 above)
Comment (AdityaKhatri) [reply src/views/Gidd/MapDashboard/index.tsx:255]: We can do that, but we also need the length of how many exist. I find this a bit easier.

File: src/views/Gidd/MapDashboard/index.tsx (hunk 9 above)
Comment (AdityaKhatri) [reply src/views/Gidd/MapDashboard/index.tsx:291]: For some reason the typing inside the if condition didn't infer when i used Array.includes

File: src/views/Gidd/MapDashboard/index.tsx (hunk 10 above)
Comment (AdityaKhatri) [reply src/views/Gidd/MapDashboard/index.tsx:316]: The paginatedData and sortedData both are exposed outside, if i didn't make a copy, the array would be mutated when I perform splice and trim.

File: src/views/Gidd/Disaster/index.tsx (hunk 11 above)
Comment (AdityaKhatri) [reply src/views/Gidd/Disaster/index.tsx:314]: For some reason the typing inside the if the condition didn't infer when I used Array.includes

File: src/Root/index.tsx [hunk 12]
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
 import { init, ErrorBoundary } from '@sentry/react';
+import 'requestidlecallback-polyfill';
```
Comment (tnagorra) [inline src/Root/index.tsx:3]: Was this added to package.json sideEffects ?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/Root/index.tsx (hunk 12 above)
Comment (AdityaKhatri) [reply src/Root/index.tsx:3]: Yes

Comment (AdityaKhatri) [reply src/Root/index.tsx:3]: No

**Review by tnagorra — APPROVED** [review]

**Pull request merged**